lib/DxcSupport/dxcapi.extval.cpp

```diff
@@ -6,37 +6,192 @@
+    _In_count_(defineCount) const DxcDefine *pDefines, // Array of defines.
+    _In_ UINT32 defineCount,                           // Number of defines.
+    _In_opt_ IDxcIncludeHandler
+        *pIncludeHandler, // User-provided interface to handle #include
+    // directives (optional).
+    _COM_Outptr_ IDxcOperationResult *
+        *ppResult // Compiler output status, buffer, and errors.
+) {
+  // First, add -Vd to the compilation args
+  UINT32 newArgCount = argCount + 1;
+  std::vector<LPCWSTR> newArgs;
+  newArgs.reserve(newArgCount);
+
+  // Copy existing args
+  for (unsigned int i = 0; i < argCount; ++i) {
+    newArgs.push_back(pArguments[i]);
+  }
+
+  // Add an extra argument
+  newArgs.push_back(L"-Vd");
+
+  return m_pCompiler->Compile(pSource, pSourceName, pEntryPoint, pTargetProfile,
+                              newArgs.data(), newArgCount, pDefines,
+                              defineCount, pIncludeHandler, ppResult);
+}
+
+// Implement the IDxcCompiler3 interface Compile command
+HRESULT STDMETHODCALLTYPE ExternalValidationHelper3::Compile(
+    _In_ const DxcBuffer *pSource, ///< Source text to compile.
+    _In_opt_count_(argCount)
+        LPCWSTR *pArguments, ///< Array of pointers to arguments.
+    _In_ UINT32 argCount,    ///< Number of arguments.
+    _In_opt_ IDxcIncludeHandler
+        *pIncludeHandler, ///< user-provided interface to handle include
+    ///< directives (optional).
+    _In_ REFIID riid,      ///< Interface ID for the result.
+    _Out_ LPVOID *ppResult ///< IDxcResult: status, buffer, and errors.
+) {
+  // First, add -Vd to the compilation args
+  UINT32 newArgCount = argCount + 1;
+  std::vector<LPCWSTR> newArgs;
+  newArgs.reserve(newArgCount);
+
+  // Copy existing args
+  for (unsigned int i = 0; i < argCount; ++i) {
+    newArgs.push_back(pArguments[i]);
+  }
+
+  // Add an extra argument
+  newArgs.push_back(L"-Vd");
+
+  return m_pCompiler3->Compile(pSource, newArgs.data(), newArgCount,
+                               pIncludeHandler, riid, ppResult);
+}
+
 HRESULT DxcDllExtValidationLoader::CreateInstanceImpl(REFCLSID clsid,
                                                       REFIID riid,
                                                       IUnknown **pResult) {
-  if (DxilExtValSupport.IsEnabled() && clsid == CLSID_DxcValidator)
-    return DxilExtValSupport.CreateInstance(clsid, riid, pResult);
+  if (pResult == nullptr)
+    return E_POINTER;
+
+  *pResult = nullptr;
+
+  // If there is intent to use an external dxil.dll
+  if (!DxilDllPath.empty() && !DxilDllFailedToLoad()) {
+    if (clsid == CLSID_DxcCompiler) {
+
+      // Create validator
+      CComPtr<IDxcValidator> m_pValidator;
+      HRESULT ValHR = DxilExtValSupport.CreateInstance<IDxcValidator>(
+          CLSID_DxcValidator, &m_pValidator);
+      if (FAILED(ValHR))
+        return ValHR;
+
+      // Wrap compiler + validator
+      // The wrapper should be selected based on the target interface
```

Comment:
I'm still a bit confused about the need for this wrapper, and whether or not we'll ever be exposing it to our users.

Regardless, I think a cleaner design here would be to have a single wrapper class that implements both interfaces.  So then we unconditionally create an instance of that class and rely on QueryInterface to hand out the appropriate interface.

Reply:
That does have a plus with simplicity, but I don't think it's straightforward to implement, because the new wrapper needs to implement the abstract functions inside IUnknown. IUnknown defines AddRef and Release, which don't provide enough information in the parameters of the function to know which member to modify (pcompiler vs pcompiler3).
I also don't like the fact that we would need 3 members, 
pValidator
pCompiler
pCompiler3,
and that in every instance of this wrapper, only 2 of the 3 members would be initialized, while the other is left uninitialized.
The way it's written now, two separate classes have the potential to be used, and their full intentional purpose is exercised.

Reply:
I think you should optimize for correctness and simplicity here.